plugins/autoconfigure/plugin.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2018-present the CoreDHCP Authors. All rights reserved
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package autoconfigure
+
+// This plugin implements RFC2563:
+// 1. If the client has been allocated an IP address, do nothing
+// 2. If the client has not been allocated an IP address
+//    (yiaddr=0.0.0.0), then:
+//    2a. If the client has requested the "AutoConfigure" option,
+//        then add the selected value and terminate processing
+//    2b. Otherwise, terminate processing and send no reply
+//
+// This plugin should be used at the end of the plugin chain,
+// after any IP address allocation has taken place.
+//
+// The optional argument is the string "DoNotAutoConfigure" or
+// "AutoConfigure" (or "0" or "1" respectively).  The default
+// is DoNotAutoConfigure.
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/coredhcp/coredhcp/handler"
+	"github.com/coredhcp/coredhcp/logger"
+	"github.com/coredhcp/coredhcp/plugins"
+	"github.com/insomniacslk/dhcp/dhcpv4"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logger.GetLogger("plugins/autoconfigure")
+
+var autoconfigure byte
+
+var Plugin = plugins.Plugin{
+	Name:   "autoconfigure",
+	Setup4: setup4,
+}
+
+var argMap = map[string]byte{
+	"0":                  0,
+	"1":                  1,
+	"DoNotAutoConfigure": 0,
+	"AutoConfigure":      1,
+}
+
+func setup4(args ...string) (handler.Handler4, error) {
+	if len(args) > 0 {
+		var ok bool
+		autoconfigure, ok = argMap[args[0]]
+		if !ok {
+			return nil, fmt.Errorf("unexpected value '%v' for autoconfigure argument", args[0])
+		}
+	}
+	if len(args) > 1 {
+		return nil, errors.New("too many arguments")
+	}
+	return Handler4, nil
+}
+
+func Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
+	if !resp.YourIPAddr.IsUnspecified() {
+		return resp, false
+	}
+
+	ac := req.Options.Get(dhcpv4.OptionAutoConfigure)
+	if ac != nil {
+		//if !bytes.Equal(ac, []byte{1}) {
+		//	log.Debugf("Unexpected value for auto-configure option from client: %v", ac)
+		//}
+		resp.UpdateOption(dhcpv4.OptGeneric(dhcpv4.OptionAutoConfigure, []byte{autoconfigure}))
```

Comment:
It seems odd that there are no other single-byte options like this already.

Do you think a type Byte (or Bool) would be the way to go, similar to Uint16 in dhcpv4/option_maximum_dhcp_message_size.go ?  Or were you thinking of a simple wrapper for OptGeneric, like `OptClientIdentifier` in dhcpv4/option_misc.go

I guess Byte is the clean way.

Reply:
Seeking input from upstream in https://github.com/insomniacslk/dhcp/pull/525

Reply:
I don't have a strong opinion on the form it should take (I don't care much about DHCPv4 in general), just mostly want to avoid using the generic types and make sure we leave most of the message abstractions in the upstream library instead of making our own in coredhcp.

fwiw if it takes a while to resolve upstream I'm perfectly fine with merging this as-is, as long as there's an open upstream PR that tracks it so we can come back and update it later